File: bindings/python/ffdas/einsum.py
from typing import overload

from ._core import _ffdas
from ._core.library import get_library_handle
from ._core.tensor_like import TensorLike, T
from ._core.tensor import reshape, empty_like


def implicit_output_modes(modes: list[str]):
    mode_count = {}
    for mode in "".join(modes):
        mode_count[mode] = mode_count.get(mode, 0) + 1

    output_modes = []
    seen = set()
    for mode in "".join(modes):
        if mode_count[mode] == 1 and mode not in seen:
            output_modes.append(mode)
            seen.add(mode)

    return "".join(output_modes)


@overload
def einsum(
    subscripts: str,
    a: T,
    b: T,
    *,
    out: None = None,
) -> T: ...
@overload
def einsum(
    subscripts: str,
    a: TensorLike,
    b: TensorLike,
    *,
    out: T,
) -> T: ...

def einsum(
    subscripts: str,
    a: TensorLike,
    b: TensorLike,
    *,
    out: TensorLike | None = None,
) -> TensorLike:
    """Binary tensor contraction using Einstein summation notation.

    Supports explicit output modes (e.g. "ij,jk->ik") and implicit mode
    where repeated indices are summed and unique indices are kept in order
    of first appearance. Both operands must have the same dtype.

    Scalar outputs (all indices contracted) are returned with shape (1,).

    Args:
        subscripts: Subscript string, e.g. "ij,jk->ik".
        a: First operand.
        b: Second operand.
        out: Pre-allocated output array.

    Returns:
        Result of the contraction.
    """
    sides = subscripts.split("->")

    if len(sides) == 1:
        modes = sides[0].split(",")
        if len(modes) != 2:
            raise ValueError(
                f"subscripts must contain exactly one ',' (got '{subscripts}')"
            )
        am, bm = modes
        outm = implicit_output_modes(modes)
    elif len(sides) == 2:
        modes = sides[0].split(",")
        if len(modes) != 2:
            raise ValueError(
                f"subscripts must contain exactly one ',' (got '{subscripts}')"
            )

        am, bm = modes
        outm = sides[1]
    else:
        raise ValueError(
            f"subscripts must contain at most one '->' (got '{subscripts}')"
        )

    dims = {m: d for m, d in zip(am, a.shape)}

    for m, d in zip(bm, b.shape):
        prev = dims.setdefault(m, d)
        if prev != d:
            raise ValueError(f"input dimensions don't match for subscript '{m}'")

    if len(set(outm) - set(dims.keys())) > 0:
        raise ValueError(f"unexpected modes in output: {set(outm) - dims.keys()}")

    scalar_output = len(outm) == 0
    if scalar_output:
        dummy_mode = chr(max(ord(m) for m in am + bm) + 1)
        outm = dummy_mode
        b = reshape(b, b.shape + (1,))
        bm = bm + dummy_mode
        dims[dummy_mode] = 1

    out_shape = tuple(dims[m] for m in outm)

    if out is None:
        out = empty_like(b, shape=out_shape)
    elif out.shape != out_shape:
        raise ValueError(
            f"invalid output shape: {tuple(out.shape)} (expected {out_shape})"
        )

    am = [ord(m) for m in am]
    bm = [ord(m) for m in bm]
    outm = [ord(m) for m in outm]

    plan = _ffdas.create_contraction(
        get_library_handle(),
        a,
        am,
        b,
        bm,
        out,
        outm,
    )

    _ffdas.contraction(
        get_library_handle(),
        plan,
        a,
        b,
        out,
    )

    if scalar_output:
        out = reshape(out, (1,))

    return out

File: bindings/matlab/+ffdas/einsum.m
function out = einsum(subscripts, a, b)
% EINSUM Binary tensor contraction using Einstein summation notation.
%   OUT = EINSUM(SUBSCRIPTS, A, B) contracts tensors A and B according to
%   the subscript string. Supports explicit output modes (e.g. 'ij,jk->ik')
%   and implicit mode where repeated indices are summed and unique indices
%   are kept in order of first appearance.
%
%   Both operands must have the same dtype. Scalar outputs (all indices
%   contracted) are squeezed to a scalar.
%
%   Examples:
%     C = ffdas.einsum('ij,jk->ik', A, B);  % matrix multiplication
%     C = ffdas.einsum('bij,bjk->bik', A, B);  % batched matmul
%     C = ffdas.einsum('i,j->ij', a, b);  % outer product
%
%   Inputs:
%     SUBSCRIPTS - Subscript string (char).
%     A          - First operand (gpuArray).
%     B          - Second operand (gpuArray).
%
%   Output:
%     OUT - Contracted result (gpuArray).

    arguments
        subscripts char
        a gpuArray  
        b gpuArray
    end
    
    sides = split(subscripts, '->');
    
    if length(sides) == 1
        modes = split(sides{1}, ',');
        if length(modes) ~= 2
            error('subscripts must contain exactly one '','' (got ''%s'')', subscripts);
        end
        am = modes{1};
        bm = modes{2};
        
        all_modes = [am, bm];
        mode_count = containers.Map();
        for i = 1:length(all_modes)
            mode = all_modes(i);
            if isKey(mode_count, mode)
                mode_count(mode) = mode_count(mode) + 1;
            else
                mode_count(mode) = 1;
            end
        end
        
        outm = '';
        seen = containers.Map();
        for i = 1:length(all_modes)
            mode = all_modes(i);
            if mode_count(mode) == 1 && ~isKey(seen, mode)
                outm = [outm, mode];
                seen(mode) = true;
            end
        end
    
    elseif length(sides) == 2
        modes = split(sides{1}, ',');
        if length(modes) ~= 2
            error('subscripts must contain exactly one '','' (got ''%s'')', subscripts);
        end
        am = modes{1};
        bm = modes{2};
        outm = sides{2};
    else
        error('subscripts must contain at most one ''->'''' (got ''%s'')', subscripts);
    end
    
    if ~strcmp(class(a), class(b))
        error('Input tensors must have the same datatype');
    end
    
    am_int = int32(am) - int32('a');
    bm_int = int32(bm) - int32('a');
    outm_int = int32(outm) - int32('a');
    
    scalar_output = isempty(outm);
    if scalar_output
        dummy_mode = max([am_int, bm_int]) + 1;
        outm_int = dummy_mode;
        
        b = reshape(b, [size(b), 1]);
        bm_int = [bm_int, dummy_mode];
    end
    
    h = ffdas.core.get_handle();

    out = ffdas.core.ffdas_contraction(h, a, am_int, b, bm_int, outm_int);
    
    if scalar_output
        out = squeeze(out);
    end
end
